### PR TITLE
enable building of ld.gold in binutils 2.25

### PR DESCRIPTION
**Pull request by ocaisa**
Add support for ld.gold, PIC build and instructions on how to include libiberty in installation

### CODE REVIEW

**Comment by ocaisa** [comment]
ld.gold is significantly faster for C++ linking (I've read)

**Comment by boegel** [comment]
Jenkins: ok to test

File: easybuild/easyconfigs/b/binutils/binutils-2.25.eb [hunk 1]
```diff
@@ -18,12 +18,19 @@ builddependencies = [('zlib', '1.2.8')]
...
+# make sure that system libraries are also considered by ld and ld.gold is also built
+configopts = '--with-sysroot=/ --enable-gold --enable-ld=default '
 
-binlist = ['addr2line', 'ar', 'as', 'c++filt', 'elfedit', 'gprof', 'ld', 'ld.bfd', 'nm',
-           'objcopy', 'objdump', 'ranlib', 'readelf', 'size', 'strings', 'strip' ]
+# Ensure PIC
+buildopts = 'CFLAGS="-fPIC"'
+
+# If you require libiberty uncomment the following
+#postinstallcmds = ["cp libiberty/libiberty.a %(installdir)s/lib/; cp include/libiberty.h %(installdir)s/include/",
```
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.25.eb]: wait, what? this is the only/proper way? there's no way to make `make install` do this for you? hard to believe...

File: easybuild/easyconfigs/b/binutils/binutils-2.25.eb [hunk 2]
```diff
@@ -18,12 +18,19 @@ builddependencies = [('zlib', '1.2.8')]
...
 
-# make sure that system libraries are also considered by ld
-configopts = '--with-sysroot=/'
+# make sure that system libraries are also considered by ld and ld.gold is also built
+configopts = '--with-sysroot=/ --enable-gold --enable-ld=default '
 
-binlist = ['addr2line', 'ar', 'as', 'c++filt', 'elfedit', 'gprof', 'ld', 'ld.bfd', 'nm',
-           'objcopy', 'objdump', 'ranlib', 'readelf', 'size', 'strings', 'strip' ]
+# Ensure PIC
+buildopts = 'CFLAGS="-fPIC"'
```
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.25.eb]: why is this needed exactly (include a comment)?

**Comment by boegel** [comment]
Can you also apply this to the other binutils 2.25 easyconfigs? That way, we'll have `ld.gold` available in the `2015b` toolchains too....

File: easybuild/easyconfigs/b/binutils/binutils-2.25.eb (hunk 2 above)
Comment (ocaisa) [inline easybuild/easyconfigs/b/binutils/binutils-2.25.eb]: I'm pulling this from memory but I think this came about as a result of a need for libiberty.a to be PIC (maybe I should just include it in the commented section?)

File: easybuild/easyconfigs/b/binutils/binutils-2.25.eb (hunk 1 above)
Comment (ocaisa) [inline easybuild/easyconfigs/b/binutils/binutils-2.25.eb]: libiberty is built by default but not included in the installation (says in configure `# Note that libiberty is not a target library`). Libiberty source is hosted in GCC and that's the correct place to enable the library but on the mailing lists people said that they might have a non-PIC GCC build

**Comment by ocaisa** [comment]
Once we agree on the changes here I'll add them to the other easyconfigs

**Comment by ocaisa** [comment]
Will I dump the libiberty commented section? Maybe it's just a site-specific customisation.

**Comment by boegel** [comment]
@ocaisa: yes, I feel the libiberty stuff should be dropped, it looks like a can of worms to me (having a two libiberty.a's around, one provided by `binutils` (built with `-fPIC`), and one provided by GCC...)

**Comment by boegel** [comment]
@wpoely86: can you look into a test report on Debian for this one?

**Comment by ocaisa** [comment]
Well, it's not provided by GCC unless you explicitly ask for it...and then it becomes connected to a toolchain, that was kind of the problem.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/68a0f7e0d59ee9a5001f for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/b9cc07ca8f2c7ad36959 for a full test report.

**Comment by boegel** [comment]
Hmm, looks troublesome on SL6:

```
script-c.h:221: error: ‘YYSTYPE’ was not declared in this scope                                                                                                                                                                                                           script-c.h:221: error: expected primary-expression before ‘,’ token                                                                                                                                                                                                        script-c.h:221: error: expected primary-expression before ‘void’   
```

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
Linux red hat enterprise linux computenode 6.5, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.6.6
See https://gist.github.com/d2261533625e96c8c019 for a full test report.

**Comment by boegel** [comment]
Not sure what to do about `libiberty`, but just including it because we vaguely remember it was needed for one particular build doesn't seem like a good idea.

Also, it seems like `libiberty.a` *is* there by default with binutils 2.22? See https://github.com/hpcugent/easybuild-easyconfigs/blob/master/easybuild/easyconfigs/b/binutils/binutils-2.22-goolf-1.4.10.eb#L32

Anyway, we'll need to figure out why the build is suddenly broken on SL6 now, see test reports that were submitted.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
Linux debian 8.1, Intel(R) Xeon(R) CPU            5130  @ 2.00GHz, Python 2.7.9
See https://gist.github.com/40fdb54256608feb5dfd for a full test report.

**Comment by wpoely86** [comment]
Make the same change in all binutils-2.25 easyconfigs?

**Comment by boegel** [comment]
build problem on SL6 looks related to `flex` and/or `Bison`, so we'll need to include these a build deps?

**Comment by boegel** [comment]
see https://sourceware.org/ml/binutils/2013-06/msg00244.html

**Comment by wpoely86** [comment]
oh great, more build dependencies. Yeah, add them as build deps.

**Comment by boegel** [comment]
tackled in https://github.com/ocaisa/easybuild-easyconfigs/pull/5

**Comment by boegel** [comment]
SL6 (no Bison in the OS):

```
 rpm -qa | egrep -i 'flex|bison'
flex-2.5.35-9.el6.x86_64
```

RHEL6:

```
$ rpm -qa | egrep -i 'flex|bison'
bison-2.4.1-5.el6.x86_64
flex-2.5.35-8.el6.x86_64
```

**Comment by wpoely86** [comment]
what? How can there be no bison?

File: easybuild/easyconfigs/b/binutils/binutils-2.25.eb [hunk 3]
```diff
@@ -12,18 +12,27 @@ sources = [SOURCE_TAR_GZ]
...
 prebuildopts = 'LIBS="$EBROOTZLIB/lib/libz.a"'
 
-# make sure that system libraries are also considered by ld
-configopts = '--with-sysroot=/'
+# make sure that system libraries are also considered by ld and ld.gold is also built
+configopts = '--with-sysroot=/ --enable-gold --enable-ld=default '
 
-binlist = ['addr2line', 'ar', 'as', 'c++filt', 'elfedit', 'gprof', 'ld', 'ld.bfd', 'nm',
-           'objcopy', 'objdump', 'ranlib', 'readelf', 'size', 'strings', 'strip' ]
+# If you require libiberty uncomment the following (you probably want a PIC build for linking purposes if you need libiberty):
```
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.25.eb]: @ocaisa: this should be dropped?

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/68447ef4cde03e9cc903 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
Linux debian 8.1, Intel(R) Xeon(R) CPU            5130  @ 2.00GHz, Python 2.7.9
See https://gist.github.com/ba49601dd4812dc69ec6 for a full test report.

File: easybuild/easyconfigs/b/binutils/binutils-2.25.eb (hunk 3 above)
Comment (ocaisa) [inline easybuild/easyconfigs/b/binutils/binutils-2.25.eb]: Reluctantly yes, libiberty used to be provided by binutils (up to 2.22 I think) though the real source is in GCC. The problem with that is it becomes a toolchain dependent library then. Our site customisation will leave this in I think.

**Comment by ocaisa** [comment]
For those who run into a libiberty issue and an online search sends you here, you can add libiberty to your binutils installations with the following:
```
# You probably want a PIC build for linking purposes if you need libiberty):
buildopts = 'CFLAGS="-fPIC"'
postinstallcmds = ["cp libiberty/libiberty.a %(installdir)s/lib/; cp include/libiberty.h %(installdir)s/include/",
                   "mkdir %(installdir)s/info",
                   "cp libiberty/libiberty.texi %(installdir)s/info/"]
-
```
Libiberty used to be included in a binutils installation but is not any longer (as of 2.23 I think), the actually source sits in GCC and can be enabled there with the configure option `--enable-install-libiberty`.

**Comment by boegel** [comment]
It seems like GCC always ship libiberty?

See https://gcc.gnu.org/ml/gcc/2013-03/msg00302.html and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=56780

**Comment by ocaisa** [comment]
Only the header file by default, not the library (at least in my GCC installations!)

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 16 out of 17 (17 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/5df4d5db36316c92541e for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 16 out of 16 (16 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/78349c2988a7c7daf988 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 6 out of 25 (16 easyconfigs in this PR)
Linux red hat enterprise linux computenode 6.5, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.6.6
See https://gist.github.com/6cf7b5d29bc7c6507683 for a full test report.

**Comment by boegel** [comment]
The latest failed test shows that we need to include binutils as a build dep in several of the added easyconfigs, to make things work on RHEL6 + Haswell, like we do in the existing `zlib` easyconfig:

```
builddependencies = [('binutils', '2.25', '', True)]
```

**Comment by boegel** [comment]
tackled in https://github.com/ocaisa/easybuild-easyconfigs/pull/10

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 16 out of 16 (16 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/484f38cd99bd6eb186d4 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 7 out of 23 (16 easyconfigs in this PR)
Linux red hat enterprise linux computenode 6.5, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.6.6
See https://gist.github.com/6eed61f3e6b56eae3294 for a full test report.

**Comment by boegel** [comment]
@ocaisa: GCC build fails on RHEL6 with:

```
ld.gold: -plugin: unknown option
ld.gold: use the --help option for usage information
collect2: error: ld returned 1 exit status
```

Either we screwed up the default `ld`, or GCC finds `ld.gold`, tries to use it, and fails...

**Comment by boegel** [comment]
Or we need `--enable-plugins`: http://stackoverflow.com/questions/23754665/gold-linker-ld-gold-plugin-unknown-option

**Comment by wpoely86** [comment]
Can't we simply not enable gold for the bootstrap binutils? The only advantage is a slightly faster linking time. I would only do it for the binutils used for actually building codes.

**Comment by boegel** [comment]
@wpoely86: sure, but we still needs `--enable-plugins` whenever we use `--enable-gold`

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 15 out of 16 (16 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/f768681999676d24a4ac for a full test report.

**Comment by ocaisa** [comment]
Typo - that will succeed now

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 16 out of 16 (16 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/b07d9ac902bb4aae1dd5 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 16 out of 16 (16 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/766e34fd4d5943cd4412 for a full test report.

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 22 out of 22 (16 easyconfigs in this PR)
Linux debian 8.1, Intel(R) Xeon(R) CPU            5130  @ 2.00GHz, Python 2.7.9
See https://gist.github.com/e6f6ba6b1539c5ae24a3 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 22 out of 26 (16 easyconfigs in this PR)
Linux red hat enterprise linux computenode 6.5, Intel(R) Xeon(R) CPU E5-2650 v3 @ 2.30GHz, Python 2.6.6
See https://gist.github.com/bece0c53cd88ecf53f62 for a full test report.

**Comment by boegel** [comment]
The builds that fail in the last test report are the ones using the default `binutils` provided by the OS.

Since this is a RHEL6 Haswell systems, the default binutils is broken; it doesn't recognise AVX2 instructions which leads to errors like:

```
Error: no such instruction: `shlx %eax,%edx,%edx'
```

In short: it's normal that these builds fail.

Thus: good to go in. Thanks @ocaisa for diving into this!

File: easybuild/easyconfigs/b/binutils/binutils-2.25.eb (hunk 1 above)
Comment (ocaisa) [inline easybuild/easyconfigs/b/binutils/binutils-2.25.eb]: This has popped up again in https://github.com/hpcugent/easybuild-easyblocks/issues/969#issuecomment-231339463

Not sure things have progressed within binutils since then, but from  this thread including libiberty from binutils is non-standard so we should maybe think hard about it.
